### PR TITLE
Add `pretty_terminator` to pretty stable-mir

### PR DESCRIPTION
**Pull request by ouz-a**
~Because we don't have successors in `stable_mir` this is somewhat lacking but it's better than nothing~, also fixed bug(?) with `Opaque` which printed extra `"` when we try to print opaqued `String`.

**Edit**: Added successors so this covers Terminators as a whole.
r? @celinval

### CODE REVIEW

**Comment by ouz-a** [comment]
> Is there a way that we could have some sort of stdout test that shows what the output changes are? Like a small, stable example program that we could have blessed every time the output changes?

I was thinking of writing one after couple PRs down the line.

File: compiler/stable_mir/src/mir/body.rs [hunk 1]
```diff
@@ -83,6 +83,7 @@ impl Body {
                         Ok(())
                     })
                     .collect::<Vec<_>>();
+                writeln!(w, "{}", pretty_terminator(&block.terminator.kind))?;
```
Comment (compiler-errors) [inline compiler/stable_mir/src/mir/body.rs:86]: Huh, so we were just not printing the terminator at all? Lol

File: compiler/stable_mir/src/mir/pretty.rs [hunk 2]
```diff
@@ -69,6 +71,131 @@ pub fn pretty_statement(statement: &StatementKind) -> String {
     pretty
 }
 
+pub fn pretty_terminator(terminator: &TerminatorKind) -> String {
```
Comment (compiler-errors) [inline compiler/stable_mir/src/mir/pretty.rs:74]: Unrelated, I wonder if it's more efficient to treat these more like the `Display`/`Debug` traits, where they write directly to a formatter. That's unrelated, though.

File: compiler/stable_mir/src/mir/pretty.rs [hunk 3]
```diff
@@ -69,6 +71,131 @@ pub fn pretty_statement(statement: &StatementKind) -> String {
...
+            );
+            pretty
+        }
+        AssertMessage::DivisionByZero(op) => {
+            let pretty_op = pretty_operand(op);
+            pretty.push_str(format!("\"attempt to divide `{{}}` by zero\", {pretty_op}").as_str());
+            pretty
+        }
+        AssertMessage::RemainderByZero(_) => {
+            format!("attempt to calculate the remainder with a divisor of zero")
```
Comment (compiler-errors) [inline compiler/stable_mir/src/mir/pretty.rs:183]: Why does this message differ from [`AssertMessage::fmt_assert_args`](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_middle/mir/syntax/enum.AssertKind.html#method.fmt_assert_args)? It should say 

> "attempt to calculate the remainder of `{{}}` with a divisor of zero", {op:?}"

Could we actually just use that method instead of duplicating these messages?

**Review by compiler-errors — CHANGES_REQUESTED** [review]

File: compiler/stable_mir/src/mir/body.rs (hunk 1 above)
Comment (ouz-a) [reply compiler/stable_mir/src/mir/body.rs:86]: We do print them

File: compiler/stable_mir/src/mir/pretty.rs (hunk 3 above)
Comment (ouz-a) [reply compiler/stable_mir/src/mir/pretty.rs:183]: Well I would have to write rustc version of `AssertKind` which transforms `stable_mir::AssertKind` to `AssertKind`, I try to avoid duplication whenever I can. Also good catch I was looking at wrong message 🤣

Comment (compiler-errors) [reply compiler/stable_mir/src/mir/pretty.rs:183]: Oh, they're different types.

File: compiler/stable_mir/src/mir/pretty.rs (hunk 2 above)
Comment (compiler-errors) [reply compiler/stable_mir/src/mir/pretty.rs:74]: I'm just saying there could be some trait like

```
trait Pretty {
  fn fmt(&self, fmt: &mut fmt::Formatter) -> fmt::Error;
}
```

Which duplicates the functionality of `Pretty`, possible even with a `context` passed as an argument, but which does the pretty-printing without having to allocate so many intermediate strings.

File: compiler/stable_mir/src/mir/body.rs (hunk 1 above)
Comment (compiler-errors) [reply compiler/stable_mir/src/mir/body.rs:86]: I'm saying previously, were they not printed?

Comment (ouz-a) [reply compiler/stable_mir/src/mir/body.rs:86]: Ah yea, we were not printing them I thought they weren't needed because we don't have successors in `stable_mir` but It looked very sparse, `stable_mir` is still living in stone age

File: compiler/stable_mir/src/mir/pretty.rs (hunk 2 above)
Comment (ouz-a) [reply compiler/stable_mir/src/mir/pretty.rs:74]: Oh we also discussed that here https://github.com/rust-lang/rust/pull/117745#issuecomment-1810177226 I'm not sure if benefits outweigh the cons

**Comment by celinval** [comment]
> Because we don't have successors in `stable_mir` this is somewhat lacking but it's better than nothing, also fixed bug(?) with `Opaque` which printed extra `"` when we try to print opaqued `String`.
> 
Can you add `successors` functions for `Terminator` and `BasicBlock` as part of this PR?

File: compiler/stable_mir/src/mir/body.rs [hunk 4]
```diff
@@ -602,9 +663,9 @@ pub struct Constant {
 }
 
 #[derive(Clone, Debug, Eq, PartialEq)]
-pub struct SwitchTarget {
-    pub value: u128,
-    pub target: usize,
+pub struct SwitchTargets {
+    pub value: Vec<u128>,
+    pub targets: Vec<usize>,
```
Comment (celinval) [inline compiler/stable_mir/src/mir/body.rs:668]: Why did you change this? I usually prefer one vector of pairs than a pair of vectors, since the first makes it explicit which values should always go together.

**Review by celinval — COMMENTED** [review]
Looks good. I just added a comment regarding the change to SwitchTarget.

It would be nice if we could add a little sanity check for stable-mir files. We could extend the `project-stable-mir` test infrastructure for now. In that case, I wouldn't do a full match though, to avoid this constantly breaking.

File: compiler/stable_mir/src/mir/body.rs (hunk 4 above)
Comment (ouz-a) [reply compiler/stable_mir/src/mir/body.rs:668]: Because we could have multiple(2) targets for one value, see here https://doc.rust-lang.org/nightly/nightly-rustc/rustc_middle/mir/struct.SwitchTargets.html, they are always one pair, and I found this to be easier to reason about.

**Comment by ouz-a** [comment]
> It would be nice if we could add a little sanity check for stable-mir files. We could extend the `project-stable-mir` test infrastructure for now. In that case, I wouldn't do a full match though, to avoid this constantly breaking.

I'm planning that for soon, in the next PR I'm going to first projection printing and place printing and then we can have some sort of a reference program where with each change we could generate new output and see how it looks.

File: compiler/stable_mir/src/mir/body.rs (hunk 4 above)
Comment (celinval) [reply compiler/stable_mir/src/mir/body.rs:668]: Gotcha, I did notice the following comment that I think we should consider here:

https://github.com/rust-lang/rust/blob/3166210d68fe43161ef64d682dd604cc60e7607f/compiler/rustc_middle/src/mir/syntax.rs#L836-L844

Comment (ouz-a) [reply compiler/stable_mir/src/mir/body.rs:668]: 🤔 would this really matter for our case, and they haven't found a case where it's better than the one they have 😄 what do you recommend ?

Comment (celinval) [reply compiler/stable_mir/src/mir/body.rs:668]: I thnk in the rust compiler case, it makes sense to avoid a big refactoring, but for our case, I think the structure suggested in the comment is cleaner and less error prone. I would prefer that unless we have a stronger reason not to.

File: compiler/rustc_smir/src/rustc_smir/convert/mir.rs [hunk 5]
```diff
@@ -578,13 +578,15 @@ impl<'tcx> Stable<'tcx> for mir::TerminatorKind<'tcx> {
...
-                        target: target.as_usize(),
-                    })
-                    .collect(),
+                targets: {
+                    let mut value_vec = Vec::new();
+                    let mut target_vec = Vec::new();
+                    targets.iter().for_each(|(value, target)| {
+                        value_vec.push(value);
+                        target_vec.push(target.as_usize());
+                    });
```
[the remark this answers is not in the document]
Comment (ouz-a) [reply compiler/rustc_smir/src/rustc_smir/convert/mir.rs]: simplified it

File: compiler/stable_mir/src/mir/body.rs (hunk 4 above)
Comment (ouz-a) [reply compiler/stable_mir/src/mir/body.rs:668]: Hmm found quite annoying bug, currently we never print `otherwise`  because we have wrong number of successors, e.g for this `[0,1,"otherwise"]` and successors count would be `3` in `stable_mir` it's `2` this might be due to fact we don't have some of the `TerminatorKind` variants in `stable_mir` 😕

Comment (celinval) [reply compiler/stable_mir/src/mir/body.rs:668]: I think the problem is with the conversion from MIR to StableMIR. We're not translating the otherwise target.

If you do prefer keeping the structure as is, can you please make the members private and provide methods to read and write them please?

File: compiler/rustc_smir/src/rustc_smir/convert/mir.rs [hunk 6]
```diff
@@ -578,13 +578,13 @@ impl<'tcx> Stable<'tcx> for mir::TerminatorKind<'tcx> {
...
-                    .iter()
-                    .map(|(value, target)| stable_mir::mir::SwitchTarget {
-                        value,
-                        target: target.as_usize(),
-                    })
-                    .collect(),
+                targets: {
+                    let (value_vec, mut target_vec): (Vec<_>, Vec<_>) =
+                        targets.iter().map(|(value, target)| (value, target.as_usize())).unzip();
+                    // We need to push otherwise as last element to ensure it's same as in MIR.
```
Comment (kw-fn) [inline compiler/rustc_smir/src/rustc_smir/convert/mir.rs:584]:
```suggestion
                    // We need to push `otherwise` as last element to ensure it's same as in MIR.
```

File: compiler/stable_mir/src/mir/body.rs (hunk 4 above)
Comment (celinval) [reply compiler/stable_mir/src/mir/body.rs:668]: As discussed offline, we can work on a better structure for SwitchTargets in a follow up PR. This change is however better than our current approach which is missing `otherwise`, so let's move ahead with this for now.

File: compiler/stable_mir/src/mir/pretty.rs (hunk 2 above)
Comment (celinval) [reply compiler/stable_mir/src/mir/pretty.rs]: I had the same thought as @compiler-errors, but I think we can leave the implementation details to a later point when we can properly assess the trade-offs. The methods here will be helpful for users, and we can change them to invoke a `fmt` function to optimize the writing process.
